Add light theme

adtac opened the pull request:
Here's what it looks like:

![light](https://user-images.githubusercontent.com/7521600/56179971-8ffe6f00-5fd5-11e9-8ad9-35b85cccfa61.png)

Closes https://github.com/browserpass/browserpass-extension/issues/98

adtac commented:
I'm not very good with frontend stuff, so please feel free to point out anything stupid lol

adtac commented:
@erayd thanks for the review. Previously, match text colour was coupled with the hint background colour for some reason. I've decoupled the two now, so it should be much better. Here's what the popup looks like now:

![Untitled](https://user-images.githubusercontent.com/7521600/56181994-fbe4d580-5fdd-11e9-8ef7-0558ecf3ab68.png)

adtac commented:
(Made the badge a bit less dark to improve text contrast.)

erayd commented:
Definitely improving - thanks for your work on this :+1: .

Can you add some colour (or better contrast) back to the domain badge in the search field? It's fading into the background a bit:

![image](https://user-images.githubusercontent.com/121486/56182504-a7396d00-6066-11e9-922f-4f3fd12addfd.png)

## Other Stuff
There is no noticable difference in the search field background colour based on whether it's focused or not. Compare with the dark theme to see what I mean - the dark theme gets this bit right.

When moving focus from the search field to the first item (e.g. by pressing the down-arrow key), the first item gets much lighter, and gives the impression that it's just been deselected. Changing the colour or shade is fine, but it needs to indicate to the user that this element has *gained* focus, not lost it.

erayd commented:
Here's another screenshot showing what it looks like with multiple custom stores configured:
![scr-20190416-165402](https://user-images.githubusercontent.com/121486/56182988-a43f7c00-6068-11e9-855f-477073f61147.png)

adtac commented:
@erayd updated. I was a bit hesitant to add non-grey colours to the mix as I didn't want it to be jarring, but I suppose it's alright to make the badge stand out a bit more.

![Untitled](https://user-images.githubusercontent.com/7521600/56183376-51bc7c00-5fe4-11e9-9ba5-8ebcf99f0019.png)

Hover is now darker than active so that it's more intuitive. The search field gets subtly lighter when it goes out of focus as well.

erayd commented:
Getting there. Now the domain badge has very low contrast with the 'remove-domain-badge' button though.

adtac commented:
Heh, you have a very keen eye, I missed that. Updated to a safer yellow :)

maximbaz commented on src/popup/colors.less:
```diff
@@ -0,0 +1,78 @@
+    .part.login > .action:focus,
+    .part.login:focus > .name {
+        background-color: @hover-bg-color;
+        outline: none;
```
I think outline needs to be declared in `popup.less`, not here

maximbaz commented on src/popup/popup.js:
```diff
@@ -117,6 +117,10 @@ async function run() {
+
+        document
+            .getElementsByTagName("html")[0]
+            .setAttribute("class", "colors-" + settings["theme"]);
```
Use ``.classList.add(`colors-${settings.theme}`)``, so that we don't accidentally override an existing class on `<html>` tag (now or in future).

maximbaz replied:
In addition, please move this snippet to an earlier stage in the function, right after we retrieve settings. Right now I see a visible change of colors, when I open the popup, at first it is something white, and then in a split second it changes to dark theme.

Managed to catch a screenshot of it:

![image](https://user-images.githubusercontent.com/1177900/56245842-b0473000-60a0-11e9-8c96-da869af6d91a.png)

I confirm that this fixes the issue for my particular case, but to be on a safe side, can we perhaps even add `class="colors-dark"` directly in `popup.html`, to keep the theme by default dark even while we are loading settings? And this snippet would need to be adjusted to something like:

```
const root = document.getElementsByTagName("html")[0];
root.classList.remove("colors-light", "colors-dark");
root.classList.add("colors-${settings.theme}");
```

maximbaz commented:
Finally got to try this! Here's some of my feedback:

1. I think we should be consistent in the way we highlight filter matches, if it is a different _text color_ in dark theme, let's use text color in light theme too, not _background color_:

    ![image](https://user-images.githubusercontent.com/1177900/56246516-78d98300-60a2-11e9-9b2e-af026d9abbe8.png)

    ![image](https://user-images.githubusercontent.com/1177900/56246527-80992780-60a2-11e9-8d63-0b9b56f545bf.png)

1. The "Copy" buttons look like they are disabled, perhaps because the icon color is too light on a light background. Do you guys have the same impression?

1. The "Copy" buttons and the "recency" badge (and maybe even the password store badge) look particularly hard to distinguish when the mouse is hovering the row:

    ![image](https://user-images.githubusercontent.com/1177900/56246697-fe5d3300-60a2-11e9-9d9c-322a62eb4939.png)

    ![image](https://user-images.githubusercontent.com/1177900/56246713-0ae18b80-60a3-11e9-8a9b-438acd1a9cd0.png)

    ![image](https://user-images.githubusercontent.com/1177900/56246740-1d5bc500-60a3-11e9-9450-d78c26a9820d.png)

    ![image](https://user-images.githubusercontent.com/1177900/56246750-28aef080-60a3-11e9-8a35-fdb443ef288c.png)

1. I agree that the domain filter should probably be light blue instead of red or yellow, a couple of suggestions just to give an impression, what do you think?

    background-color: #aad2f5;
    color: #343a40;

    ![image](https://user-images.githubusercontent.com/1177900/56247074-f356d280-60a3-11e9-99a4-aeeb72951a24.png)

    background-color: #39a3ff;
    color: white;

    ![image](https://user-images.githubusercontent.com/1177900/56247103-0cf81a00-60a4-11e9-9a48-27b12a0af699.png)

maximbaz commented:
/cc a couple of people who I recall to be passionate about design and everyone who upvoted #98 so far, just to get a couple of extra opinions: @DamienCassou, @532910, @JanMa, @jakubtopic, @Tharre.

If anyone can't compile the branch to test, I'm willing to provide test builds, just say so.

I'd very much prefer to spend an extra time and get this right on the first try, so please, early feedback is very welcome.

Tharre commented:
The buttons look grayed out/disabled indeed.

The darker blue looks the nicest to me, but you might want to tone down the red if you go with it.

erayd commented:
> I think we should be consistent in the way we highlight filter matches, if it is a different *text color* in dark theme, let's use text color in light theme too, not *background color*.

I disagree with this point - I actually think the way he's done the highlighting is spot on. Yes, it's different, but I don't see any issue with different themes being different, and the downside of a light pallette is it's harder to get good contrast with a text-colour change, the way the dark theme does it.

adtac commented:
>I think we should be consistent in the way we highlight filter matches, if it is a different text color in dark theme, let's use text color in light theme too, not background color:

It's currently both a different text colour *and* a different background colour. And as @erayd said, I personally think a small background contrast adds a lot to the visibility of the search matches. In fact, I wanted to update the dark theme to support this as well (I can still do this if others like the light theme's way of doing search matches better).

>The "Copy" buttons look like they are disabled, perhaps because the icon color is too light on a light background. Do you guys have the same impression?

>The "Copy" buttons and the "recency" badge (and maybe even the password store badge) look particularly hard to distinguish when the mouse is hovering the row

Agreed, I've updated this. It was a bit tricky because the classes use a SVG background with a hard-coded fill colour. I didn't want to create a separate SVG for the light theme as that's just an unnecessary maintenance overhead. As a result, the light theme is using CSS invert. While this is a perfectly valid solution for light/dark themes, implementing other themes will likely require hue rotate and other shenanigans.

I've also updated the badge to a dark blue. I agree with @Tharre that it looks nicer.

>I confirm that this fixes the issue for my particular case, but to be on a safe side, can we perhaps even add class="colors-dark" directly in popup.html, to keep the theme by default dark even while we are loading settings? And this snippet would need to be adjusted to something like

If the dark theme is the default, I agree. I've made the change you suggested, but I'd also like to start a discussion on what the default theme itself should be. I think there's a case to be made that the default theme should be light for two reasons:

 - Most browsers come with a default light theme, so it's a bit jarring to have an extension not follow the standard.
 - v2 used to have a light theme, so sticking to a light theme would make the upgrade easier from a muscle memory perspective.

JanMa commented:
So far I like the light theme a lot. Great work @adtac :-)
Maybe you could think about using the same yellow colours for highlighting as in the dark theme but this is just personal preference. I would also favour the light theme to be the default one, because light themes are also the default in Chrome and Firefox.

DamienCassou commented:
I have used the light theme the whole day and I like it. Thank you for your work.

maximbaz commented:
Getting much better!

1. I couldn't leave without trying to highlight with text color alone (`#00bcff`), IMHO looks quite alright, is it worse in your opinion than the highlight with background color?

    ![image](https://user-images.githubusercontent.com/1177900/56325703-74c96600-6173-11e9-867c-8ea4e3510e46.png)


1. If you still think the background highlight is much better, could it be made so that the blue background doesn't "touch" the store badge? In the picture below, ideally the blue background would stop maybe just a bit above the top of the "h" letter, but I feel there should be some space left between blue background and the store badge.

    ![image](https://user-images.githubusercontent.com/1177900/56325246-ca047800-6171-11e9-915f-d83760530ce5.png)

1. Would you agree that there's not enough contrast in this case, between blue background and the background of this entry? It doesn't look as bad on the enlarged screenshot, better try this in your browser:

    ![image](https://user-images.githubusercontent.com/1177900/56325359-3aab9480-6172-11e9-8fd6-5ba389f2ba43.png)

1. When an entry is hovered, it feels the store badge also has small contrast (again doesn't look that bad on the enlarged screenshot, but try in your browser):

    ![image](https://user-images.githubusercontent.com/1177900/56325447-852d1100-6172-11e9-9c6d-fb551c16ce2e.png)

erayd commented:
@maximbaz I still think the background highlight is better than the change in text color - the  text in your example fades into the background somewhat. I think the background highlight colour is fine as-is, and the contrast is sufficient. Remember that it contrasts with the text too, not just the background.

I agree that there needs to be space between the highlight and the badge. And perhaps the badge colour could change a little too improve contrast in your point (4).

adtac commented:
> 1. I couldn't leave without trying to highlight with text color alone (`#00bcff`), IMHO looks quite alright, is it worse in your opinion than the highlight with background color?

Not at all. I still think the light background makes it stand out a bit more. However, if you feel strongly about this, please don't hesitate to ask me to implement it without the background; it's your project, after all :) I don't mind either, but I prefer the background (and apparently, so do some others).

> 2. If you still think the background highlight is much better, could it be made so that the blue background doesn't "touch" the store badge? In the picture below, ideally the blue background would stop maybe just a bit above the top of the "h" letter, but I feel there should be some space left between blue background and the store badge.

Added a 2px margin between the lines. Does not affect the height of each entry.

> 3. Would you agree that there's not enough contrast in this case, between blue background and the background of this entry? It doesn't look as bad on the enlarged screenshot, better try this in your browser:

True, made it a bit darker.

> 4. When an entry is hovered, it feels the store badge also has small contrast (again doesn't look that bad on the enlarged screenshot, but try in your browser):

Fixed as well.

> 5. I'd recommend `invert(85%)` instead of `100%`, makes the icons look not as bold:

Done (although this makes the hover background logic quite tricky as we now need to use a colour such that `invert(85%)` is the same as `@hover-bg-color`, for which only an approximate solution exists). We could change the SVG's colour slightly so that `invert(100%)` isn't so contrasting, but this is a less "elegant" solution IMO.

> 6. The contrast between a focused search box and a pre-selected first line might also need additional improvement:

Fixed.

> 7. Is the "search icon" on purpose left without the "invert"? On the one hand it's not a button, on the other hand in the dark theme it has the same color as the buttons, and some might say it's almost invisible due to the low contrast with the background of the focused search input box:

This is intentional. Due to `invert` and the way the search button's CSS is written, this is impossible because applying a filter would make the search background dark.

The most elegant solution (for points 5, 7) would be to inline the SVGs in the LESS and use a variable-based fill colour inside the SVG string. Perhaps this could be done in a future improvement (unfortunately, I don't have the time to do this).

BTW, please feel free to make changes as you see fit once you merge this. I think that'd be a much more productive way than having back-and-forth replies about subtle colour changes :)

erayd commented:
> Added a 2px margin between the lines. Does not affect the height of each entry.
The entry height is the same (53px), but it's stuffed up the vertical height somehow. The light theme now pushes the sixth entry off the bottom:

![image](https://user-images.githubusercontent.com/121486/56334788-ec3bdb80-61ed-11e9-94da-b110e95918e4.png)
![image](https://user-images.githubusercontent.com/121486/56334847-22795b00-61ee-11e9-8008-3baaf9cfff9f.png)

Ideally, the bottom of the popup would be flush with the bottom of the sixth entry. I'm not sure why the bottom of the popup is collapsing this way, but it needs resolving. The dark theme also exhibits this a little, but only by a very small amount.

> Done (although this makes the hover background logic quite tricky as we now need to use a colour such that invert(85%) is the same as @hover-bg-color, for which only an approximate solution exists).
Can you explain why this complexity is needed? What are you trying to achieve by having them be the same? From a visual perspective what you've done here looks fine.

> This is intentional. Due to invert and the way the search button's CSS is written, this is impossible because applying a filter would make the search background dark.
I'm actually OK with this as it is. It makes sense for the search button to be pale, because it isn't a button, it's just a hint. I don't think we need to change it.

> BTW, please feel free to make changes as you see fit once you merge this. I think that'd be a much more productive way than having back-and-forth replies about subtle colour changes :)
I disagree. Much of the point of the PR process is to get things right the first time, and merge it when it's ready. There's no deadline that we have to merge this feature by, so we can take our time here and get it right. You've done a great job here - now we're just at the nitpicking stage trying to get it exactly right.

maximbaz commented:
Awesome work, I'll give it another go later tonight 😉 

> The entry height is the same (53px), but it's stuffed up the overall vertical height of the popup somehow. The light theme now pushes the sixth entry off the bottom

I actually don't reproduce this, even if I set custom zoom in Chromium settings.

> > This is intentional. Due to invert and the way the search button's CSS is written, this is impossible because applying a filter would make the search background dark.
> 
> I'm actually OK with this as it is. It makes sense for the search button to be pale, because it isn't a button, it's just a hint. I don't think we need to change it.

Me too, just wanted to point this out to make sure it wasn't omitted 🙂

erayd commented:
> I actually don't reproduce this, even if I set custom zoom in Chromium settings.

Then let's ignore it; it's likely something funky with my zoom / setup. Chrome certainly has a number of issues in that area.

maximbaz commented:
We can ignore it for now, but it would be a pity to degrade your experience, even though you run a custom setup. Let's return to this in the end, maybe there are some other ways to implement the margin that would not cause misalignment for you...

maximbaz commented:
With the current colors I'm satisfied with everything except the default password store badge color, it has a bit of "violet" flavor that imho doesn't fit with the rest of the popup.

Status quo:

![image](https://user-images.githubusercontent.com/1177900/56421334-4d24eb80-62a2-11e9-815a-0bf109538393.png)

Let's explore other colors?

My proposals are light blue `#a9caec`:

![image](https://user-images.githubusercontent.com/1177900/56421319-354d6780-62a2-11e9-9ccc-c92c0987e0a7.png)

Or dark grey `#b2b2b2`:

![image](https://user-images.githubusercontent.com/1177900/56421328-426a5680-62a2-11e9-83d8-fcfd4590a621.png)

Opinions, other suggestions?

maximbaz commented:
I've been using the light theme for a few hours, all in all it was quite a good experience, besides the password store badge color there was one other particular scenario that threw me completely off: when I had to press <kbd>Backspace</kbd> and search manually for credentials that don't belong to the current domain, I was genuinely distracted by the highlight of matches, I think they stand out too much than necessary.

Let's take this fake example: I want to find a password "github.com" in "personal" folder, so I know that typing "pegit" would be enough to bring it to the top - and it did, but all these blue spots made me feel distracted, I felt like I am asked to pay extra attention to them all:

![image](https://user-images.githubusercontent.com/1177900/56442833-6190d480-62f2-11e9-9830-eff5353009d9.png)

I honestly didn't plan to bring up the topic of highlighting using background color again, so instead of forcing you into changing to use text color alone, let me just ask you this: could you please try this scenario for yourself and tell me if you also feel that it is distracting when the number of blue spots is big?

532910 commented:
![image](https://user-images.githubusercontent.com/14617699/56443129-e2a09980-62fb-11e9-942c-12039c25f167.png)

The one of a good design principles is to use as few colors/fonts as possible.

Why not to use #027b83 (yellow) instead of #d79921 (dark green)?
And same for light theme and blue color.

maximbaz commented:
Green looks worse on the dark grey background, and I'm so very fond of the combination of grey and orange, I don't want to change the dark theme 🙂

erayd commented:
> Let's explore other colors?
> My proposals are light blue `#a9caec`

Good call. My preference is for the blue you've proposed.

> Could you please try this scenario for yourself and tell me if you also feel that it is distracting when the number of blue spots is big?

I do find it a little distracting in this scenario. I rarely use full-store search though (and I suspect that will be the case for most people), so IMO it makes the most sense to optimise for the majority case of filtering the domain-matched list. My preference for the latter case is to keep the current highlighting approach, but I'm happy to leave that decision up to you guys.

adtac commented:
(Busy with an exam for the next couple of days, just thought I should let everybody know. I'll make amendments and push changes once I'm done.)

adtac commented:
>Or dark grey #b2b2b2

I went with a slightly lighter version to add more contrast with the "pass" text (which is black). IMO the light blue looks a bit out of place in the grey background. Lemme know what you guys think.

>I honestly didn't plan to bring up the topic of highlighting using background color again, so instead of forcing you into changing to use text color alone, let me just ask you this: could you please try this scenario for yourself and tell me if you also feel that it is distracting when the number of blue spots is big?

I agree, but as @erayd said I don't think full store searches are that common. But it's still a valid case neverthenonetheless, so I've made the colour be a bit lighter and less contrasting with the background so that it doesn't standout. I went with `#deeaf3`, but please feel free to suggest alternatives. I still think a less contrasting background is better than no background :)

Rebased on master.

adtac commented:
![Untitled](https://user-images.githubusercontent.com/7521600/57051751-ff6c9380-6c50-11e9-8ffc-9657fb6c7b0d.png)

what it looks like now ^

erayd commented:
@adtac I think I prefer something a little stronger than that for the highlight colour - I've had a bit of a play, and `#cfecff` feels about right for me. It's not as bold as the original highlight colour, but it has a bit more contrast than `#deeaf3`.

@maximbaz Thoughts?

adtac commented:
>`#cfecff`

Good call, it's a decent compromise, I agree. Updated.

maximbaz commented:
I'll try later as well 👍 

By the way, @adtac would you mind if we keep this PR open until after I finish with #61? It will make it easier for me to finish with that quite big feature if I don't have to think about two themes, and then we rebase this PR and continue with the two themes here.

adtac commented:
No problem. Just ping me here when that is merged and you're good to go.

adtac commented:
Hey @maximbaz, I haven't been following the project's development much. Any updates on the password generation feature that you wanted merged before this? I'd really like to merge this; looking at the issue page, I imagine a few others would like this too.

maximbaz commented:
Hey, I apologize for the delay and the blocked state, I'll try to prioritize that work in the following days, I would also really like to get back to zero pending PRs in this project 🙂

adtac commented:
@maximbaz ping :)

P.S. If you think the password generation feature will take more time than originally planned, I propose merging this patch first; just a suggestion :)

maximbaz commented:
Apologies again. I agree, at this point let's just get this merged, #61 is taking way too much time than I planned. @erayd could you try again if you like the current colors, as well as do the code review? I'll do the same now.

maximbaz commented:
For the background of the first entry highlighted by default I propose `#d8dee3`, a bit darker than the current `#dee2e6`, to add a bit of contrast to the blue highlight. Other than that, I'm ok with how it looks. @erayd?

erayd commented:
@maximbaz I will go over this PR tomorrow.

DamienCassou commented:
Hi guys. What is the status of this PR?

erayd commented:
@DamienCassou I'm midway through reviewing it.

DamienCassou commented:
Erayd <notifications@github.com> writes:
> @DamienCassou I'm midway through reviewing it.

thank you very much!

-- 
Damien Cassou
http://damiencassou.seasidehosting.st

"Success is the ability to go from one failure to another without
losing enthusiasm." --Winston Churchill

maximbaz commented:
Thanks @adtac for your work!

erayd commented:
Yes, thank you @adtac, and sorry we took so long!

adtac commented:
thank you!
